sycl/source/detail/device_image_impl.hpp

```diff
@@ -0,0 +1,160 @@
+//==------- device_image_impl.hpp - SYCL device_image_impl -----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/pi.h>
+#include <CL/sycl/device.hpp>
+#include <CL/sycl/kernel_bundle.hpp>
+#include <detail/device_impl.hpp>
+#include <detail/kernel_id_impl.hpp>
+#include <detail/program_manager/program_manager.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// Used for sorting vector of kernel_id's
+struct LessByNameComp {
+  bool operator()(const sycl::kernel_id &LHS, const sycl::kernel_id &RHS) {
+    return std::strcmp(LHS.get_name(), RHS.get_name()) < 0;
+  }
+};
+
+// The class is impl counterpart for sycl::device_image
+// It can represent a program in different states, kernel_id's it has and state
+// of specialization constants for it
+class device_image_impl {
+public:
+  device_image_impl(RTDeviceBinaryImage *BinImage, context Context,
+                    std::vector<device> Devices, bundle_state State)
+      : MBinImage(BinImage), MContext(std::move(Context)),
+        MDevices(std::move(Devices)), MState(State) {
+
+    // Collect kernel names for the image
+    pi_device_binary DevBin =
+        const_cast<pi_device_binary>(&BinImage->getRawData());
+    for (_pi_offload_entry EntriesIt = DevBin->EntriesBegin;
+         EntriesIt != DevBin->EntriesEnd; ++EntriesIt) {
+
+      std::shared_ptr<detail::kernel_id_impl> KernleIDImpl =
+          std::make_shared<detail::kernel_id_impl>(EntriesIt->name);
+
+      sycl::kernel_id KernelID =
+          detail::createSyclObjFromImpl<sycl::kernel_id>(KernleIDImpl);
+
+      // Insert new element keeping MKernelIDs sorted.
+      auto It = std::lower_bound(MKernelIDs.begin(), MKernelIDs.end(), KernelID,
```

Comment:
Minor: I think pushing all the elements into the vector and then sorting it once should be more efficient than inserting each one while keeping the vector sorted.

Reply:
Will do measurements.

Reply:
The problem is that it might depends a lot on the real number of kernels.
Perhaps with SYCL generated by ML frameworks could have millions of kernels... ;-)
But for "normal" code and a few kernels?
Anyway, the `std::sort` path should be fast enough too for just a few kernels...